### ISSUE TITLE
kubeflow bundle give an error multiple times during a deployment

### ISSUE THREAD

**Issue by nobuto-m** [canonical/bundle-kubeflow#458]
Following the quick installation guide:
https://charmed-kubeflow.io/docs/quickstart

A deployment settles down eventually indeed, but the `watch -c juju status --color` step reveals an error in an user visible way. It would be nice if each charm handles status appropriately like `waiting` or `blocked` instead of `error` so we can avoid an awkward moment of seeing an error in a demo.

Steps to reproduce:
1. prepare Juju with microk8s and bootstrap

       $ juju bootstrap microk8s

1. deploy kubeflow bunlde

       $ juju deploy --trust kubeflow

1. run juju-wait to monitor the deployment

       $ until time juju-wait -w; do sleep 1; done

Errors:

```
$ until time juju-wait -w; do sleep 1; done
ERROR:root:kfp-persistence/0 failed: workload status is error

real    18m11.800s
user    0m47.167s
sys     0m18.115s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    0m1.583s
user    0m0.803s
sys     0m0.230s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    0m1.593s
user    0m0.827s
sys     0m0.223s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    0m1.621s
user    0m0.776s
sys     0m0.224s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    0m1.516s
user    0m0.802s
sys     0m0.194s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    0m1.540s
user    0m0.815s
sys     0m0.237s
INFO:root:All units idle since 2022-06-05 03:38:25.777111Z (admission-webhook/0, argo-controller/0, argo-server/0, dex-auth/0, envoy/0, istio-ingressgateway/0, istio-pilot/0, jupyter-controller/0, jupyter-ui/0, katib-controller/0, katib-db-manager/0, katib-db/0, katib-ui/0, kfp-api/0, kfp-db/0, kfp-persistence/0, kfp-profile-controller/0, kfp-schedwf/0, kfp-ui/0, kfp-viewer/0, kfp-viz/0, kubeflow-dashboard/0, kubeflow-profiles/0, kubeflow-roles/0, kubeflow-volumes/0, metacontroller-operator/0, minio/0, mlmd/0, oidc-gatekeeper/0, seldon-controller-manager/0, tensorboard-controller/0, tensorboards-web-app/0, training-operator/0)

real    2m11.662s
user    0m6.220s
sys     0m2.357s
```

```
$ juju show-status-log kfp-persistence/0 --days 1
Time                   Type       Status       Message
05 Jun 2022 03:14:05Z  workload   waiting      installing agent
05 Jun 2022 03:14:05Z  juju-unit  allocating   
05 Jun 2022 03:15:04Z  workload   waiting      agent initializing
05 Jun 2022 03:15:09Z  workload   maintenance  installing charm software
05 Jun 2022 03:15:09Z  juju-unit  executing    running install hook
05 Jun 2022 03:15:17Z  juju-unit  executing    running leader-elected hook
05 Jun 2022 03:15:18Z  juju-unit  executing    running config-changed hook
05 Jun 2022 03:15:19Z  workload   blocked      Missing required relation for kfp-api
05 Jun 2022 03:15:20Z  juju-unit  executing    running start hook
05 Jun 2022 03:15:21Z  juju-unit  idle         
05 Jun 2022 03:16:10Z  juju-unit  executing    running kfp-api-relation-created hook
05 Jun 2022 03:16:12Z  juju-unit  idle         
05 Jun 2022 03:16:12Z  juju-unit  executing    running kfp-api-relation-joined hook for kfp-api/0
05 Jun 2022 03:16:13Z  juju-unit  executing    running kfp-api-relation-changed hook for kfp-api/0
05 Jun 2022 03:16:15Z  workload   waiting      List of kfp-api versions not found for apps: kfp-api
05 Jun 2022 03:16:16Z  juju-unit  idle         
05 Jun 2022 03:16:21Z  juju-unit  executing    running kfp-api-relation-changed hook
05 Jun 2022 03:16:22Z  workload   waiting      Waiting for kfp-api relation data
05 Jun 2022 03:16:23Z  juju-unit  idle         
05 Jun 2022 03:17:50Z  juju-unit  executing    running kfp-api-relation-changed hook
05 Jun 2022 03:17:52Z  workload   maintenance  Setting pod spec
05 Jun 2022 03:17:52Z  workload   active       
05 Jun 2022 03:17:53Z  juju-unit  idle         
05 Jun 2022 03:18:08Z  juju-unit  executing    running kfp-api-relation-changed hook for kfp-api/0
05 Jun 2022 03:18:08Z  workload   waiting      waiting for container
05 Jun 2022 03:18:10Z  workload   maintenance  Setting pod spec
05 Jun 2022 03:34:16Z  juju-unit  idle         
05 Jun 2022 03:34:16Z  workload   active       
05 Jun 2022 03:36:17Z  juju-unit  error        container error: 
05 Jun 2022 03:36:17Z  workload   error        container error: 
05 Jun 2022 03:36:32Z  juju-unit  idle         
05 Jun 2022 03:36:32Z  workload   active       
```
[kfp-persistence-0.log](https://github.com/canonical/bundle-kubeflow/files/8839371/kfp-persistence-0.log)

Environment:
```
$ snap version
snap    2.55.5
snapd   2.55.5
series  16
ubuntu  20.04
kernel  5.4.0-113-generic

$ microk8s kubectl version --short
Client Version: v1.21.12-3+6937f71915b56b
Server Version: v1.21.12-3+6937f71915b56b

$  juju version
4.9.31-ubuntu-amd64
```

**Comment by nobuto-m** [comment]
Another instance of deployment:

```
Deploy of bundle completed.

real    5m52.214s
user    0m0.402s
sys     0m0.220s

until time juju-wait -w; do sleep 1; done
ERROR:root:katib-ui/0 failed: workload status is error

real    0m6.938s
user    0m1.131s
sys     0m0.572s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m6.132s
user    0m0.921s
sys     0m0.260s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m1.346s
user    0m0.634s
sys     0m0.263s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m1.403s
user    0m0.733s
sys     0m0.232s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m6.529s
user    0m0.959s
sys     0m0.267s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m1.524s
user    0m0.774s
sys     0m0.201s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m1.384s
user    0m0.709s
sys     0m0.198s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m1.396s
user    0m0.749s
sys     0m0.191s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m1.428s
user    0m0.760s
sys     0m0.178s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m1.322s
user    0m0.715s
sys     0m0.175s
ERROR:root:katib-ui/0 failed: workload status is error

real    0m1.445s
user    0m0.773s
sys     0m0.184s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    20m4.068s
user    0m50.817s
sys     0m19.210s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    0m1.612s
user    0m0.817s
sys     0m0.221s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    0m1.533s
user    0m0.788s
sys     0m0.178s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    0m1.458s
user    0m0.700s
sys     0m0.251s
ERROR:root:kfp-persistence/0 failed: workload status is error

real    0m1.487s
user    0m0.763s
sys     0m0.176s
INFO:root:All units idle since 2022-06-05 11:54:49.311477Z (admission-webhook/0, argo-controller/0, argo-server/0, dex-auth/0, envoy/0, istio-ingressgateway/0, istio-pilot/0, jupyter-controller/0, jupyter-ui/0, katib-controller/0, katib-db-manager/0, katib-db/0, katib-ui/0, kfp-api/0, kfp-db/0, kfp-persistence/0, kfp-profile-controller/0, kfp-schedwf/0, kfp-ui/0, kfp-viewer/0, kfp-viz/0, kubeflow-dashboard/0, kubeflow-profiles/0, kubeflow-roles/0, kubeflow-volumes/0, metacontroller-operator/0, minio/0, mlmd/0, oidc-gatekeeper/0, seldon-controller-manager/0, tensorboard-controller/0, tensorboards-web-app/0, training-operator/0)

real    1m52.286s
user    0m5.541s
sys     0m1.907s
```

**Comment by nobuto-m** [comment]
I see it's clarified in the doc as follows:

https://charmed-kubeflow.io/docs/quickstart
> watch -c juju status --color
> 
> This will show a list of the applications and their current status. Don’t be surprised if a few show up error messages to begin with - a lot of the components rely on the operation of others, so it can take some time before everything is ready and talking to one another.

It would be nice if we could use `waiting` status for that kind of conditions instead of an error.

https://juju.is/docs/olm/status-values#heading--workload--charm-status
> waiting 	The charm is alive and well, and is waiting for some event to occur. No human intervention required.

**Comment by ca-scribner** [comment]
I tried to reproduce this and got slightly different results.  In my case, I had `dex` go into error during startup, plus several charms go into `Blocked` (which is also undesirable, as `Blocked` is described as something that should require user intervention as well).  After time, everything went to `Active`

I think the cause is a mix of race conditions and design decisions.  When there are relations, at times one side of the relation has provided its information before the other.  This is tricky because it could be appropriately labelled as `Waiting` or `Blocked`:
* the lagging side is working and will eventually send the expected data -->`Waiting` is appropriate as it will resolve itself in time
* the lagging side is broken (bad charm code, sent bad data, wrong relation, etc) and never sends the expected data --> needs human intervention, so `Blocked` or `Error` make more sense here, although it might be hard to decide which is best given the info a charm has

In the past, we decided to prefer `Waiting` here for situations that are very likely going to work out, but `Blocked` in more ambiguous ones.  This is seen in several of the KFP charms, istio-gateway, and others.  In some cases, we're probably labelling a configuration error (which ideally would be an `Error`) as `Blocked` purely for practicality reasons. 

For those that went into `Error`, (`dex`, `katib` (as @nobuto-m cited), etc), these might be the same race conditions just labelled in a different way.  We should look into these and handle them better.

Even if we remove the `Error` cases, `Blocked` also is not very satisfying to show here for transient situations, but I don't know what we could do.  Only idea that comes to mind is to set `Waiting` with a time limit, eg `if X happens for <5 minutes, status = Waiting; else status = Blocked or Error`.  That would be great, but would take some extra tooling (maybe we write a package to handle this and use stored state to remember our timings?)
